Remove reliance on .sh script

jonasgrunert opened the pull request:
First of all I want to say, that all of this is very impressive and I am very happy, that such a cool project exists.

I had some problems getting the project of the ground on my Windows system, which had to do with the bash script user in the start script. What I did was just a small hack, which replaces the bash script with a node script, which achieves the same thing.  
This is mainly used for me, but maybe wone would want to integrate it nonetheless. I am open for further questions or anything, which may be required to get this to merge.

Have a great day.

jonasgrunert committed:
Replacing bash file with node

This should enable cross-plattform development easily, while not requiring any special setups.

jonasgrunert committed:
Merge branch 'master' of https://github.com/jonasgrunert/svelte-nodegui-starter

shirakaba commented:
Thank you very much for this (and your kind words)! Really glad to receive some expertise to solve Windows compatibility :)

I've just tried it out on macOS:

```
(node:46437) UnhandledPromiseRejectionWarning: Error: spawn node_modules/.bin/qode dist/index.js ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:268:19)
    at onErrorNT (internal/child_process.js:468:16)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:46437) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:46437) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

I rewrote it as this to pass the arguments along differently and it was happy on macOS:

```js
// I hope you don't mind – I've reworded this variable
const terminationSignals = ["close", "exit", ""];

function startSubprocess() {
  return new Promise((res, rej) => {
    // On macOS, child_process strictly wants the command and args to be separated
    const [command, ...otherArgs] = process.argv.slice(2).map(normalize);
    const proc = spawn(
      command,
      otherArgs,
      {
        stdio: ["pipe", "pipe", "pipe"],
        shell:
          process.platform === "win32" ||
          /^(msys|cygwin)$/.test(process.env.OSTYPE),
      }
    );
    proc.stdout.pipe(process.stdout);
    proc.stderr.pipe(process.stderr);
    process.stdin.pipe(proc.stdin);
    // I've also expanded "e" to "event" as it's easy to conflate with "error"
    terminationSignals.forEach((event) => proc.on(event, res));
    proc.on("error", rej);
  });
}
```

I confirmed that this script exits and relaunches the app upon a HMR signal, and also ensures that the user can still read stdout and stderr from the app. If the app exits due to some other signal (such as pressing Ctrl + C to terminate the livereload script, or the signal raised by exiting the app via the quit action from Qt's menu), the live reload script terminates as expected.

So, no regressions, and fully functioning! If you could test whether my changes work on Windows too, then I'd be happy to merge them once applied!

shirakaba commented on livereload.js:
```diff
@@ -0,0 +1,28 @@
+const { spawn } = require("child_process");
+const { normalize } = require("path");
+
+const termiateSignals = ["close", "exit", ""];
```
Could you explain when we might receive a "" signal?

Should we also handle a "[disconnect](https://nodejs.org/api/child_process.html#child_process_event_disconnect)" signal?

shirakaba commented on livereload.js:
```diff
@@ -0,0 +1,28 @@
+
+function startSubprocess() {
+  return new Promise((res, rej) => {
+    const proc = spawn(process.argv.slice(2).map(normalize).join(" "), {
```
On macOS, child_process strictly wants the command and args to be separated; I got an ENOENT error when I ran the original script:

```
(node:46437) UnhandledPromiseRejectionWarning: Error: spawn node_modules/.bin/qode dist/index.js ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:268:19)
    at onErrorNT (internal/child_process.js:468:16)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:46437) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:46437) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

Could you therefore change the spawn invocation to this, which I've found does work on macOS:

```js
    const [command, ...args] = process.argv.slice(2).map(normalize);
    const proc = spawn(
      command,
      args,
      {
        stdio: ["pipe", "pipe", "pipe"],
        shell:
          process.platform === "win32" ||
          /^(msys|cygwin)$/.test(process.env.OSTYPE),
      }
    );
```

jonasgrunert replied:
This should be done. I took the liberty to only normalize the command as referenced [here](https://github.com/kentcdodds/cross-env/blob/3edefc7b450fe273655664f902fd03d9712177fe/src/command.js#L29) in cross-env.

shirakaba replied:
I think that's saying that it shouldn't be done on command argument without care. If the command arguments include any paths, each of those paths will still need to be normalised. The PR comment shows that it would break any non-path args passed in, like URLs. So we need to apply it with care – I don't have a good solution for this apart from examining the CLI args parser from Node itself (which is the same as qode) and maybe doing something with that. Otherwise, we can just normalise the first arg, which we know is a path, and leave the rest as-is (as if they are ever introduced, it's because the user has edited the template, and at that point, it's their problem to solve!).

shirakaba commented on livereload.js:
```diff
@@ -0,0 +1,28 @@
+const { spawn } = require("child_process");
+const { normalize } = require("path");
+
+const termiateSignals = ["close", "exit", ""];
```
Could you please reword this variable to `terminationSignals`.

Could you explain when we might receive a "" signal?

Should we also handle a "[disconnect](https://nodejs.org/api/child_process.html#child_process_event_disconnect)" signal?

jonasgrunert replied:
The "" signal, was just a leftover as I first added the "disconnect" event. Then I figured out we do not need it, as it only happens, when disconnect is explicitly called. The renaming of the variable makes sense and is done,

shirakaba commented on livereload.js:
```diff
@@ -0,0 +1,28 @@
+    proc.stdout.pipe(process.stdout);
+    proc.stderr.pipe(process.stderr);
+    process.stdin.pipe(proc.stdin);
+    termiateSignals.forEach((e) => proc.on(e, res));
```
Could this `e` please be reworded to `event`? Just to distinguish it from the word "error".

jonasgrunert replied:
Good idea.

shirakaba reviewed (changes requested):
Thank you very much for this (and your kind words)! Really glad to receive some expertise to solve Windows compatibility :)

I confirmed that (following my requested changes to keep macOS happy) this script exits and relaunches the app upon a HMR signal, and also ensures that the user can still read stdout and stderr from the app. If the app exits due to some other signal (such as pressing Ctrl + C to terminate the livereload script, or the signal raised by exiting the app via the quit action from Qt's menu), the live reload script terminates as expected.

So, no regressions, and fully functioning! If you could test whether my changes work on Windows too, then I'd be happy to merge them once applied!

jonasgrunert committed:
Adding required changes

Except for the handling of the disconnect signal all changes were valid and prove to work on windows.

jonasgrunert commented:
Thank you for being so fast on the review. I double checked and all changes should be implemented by now.

jonasgrunert committed:
Minor readability changes

Forgot an s on line 7 and therefore lne 21.
A friend pointed out that the "pipe" array is the default, but "inherit" automatically uses the parents process stdio.

shirakaba commented on livereload.js:
```diff
@@ -4,21 +4,19 @@ const { normalize } = require("path");
 function startSubprocess() {
   return new Promise((res, rej) => {
     const [command, ...args] = process.argv.slice(2);
     const proc = spawn(normalize(command), args, {
```
There's no ideal way to approach this without building a full parser for `qode`'s args (technically we could take this straight out of the node library itself, as it's the same CLI interface), but at the very least, we'll need to normalise the first argument, `./dist/index.js` of the `args` array, as it also contains a path.

Here's my outdated review comment about `cross-env` saying that only the command should be normalised:

> I think that's saying that it shouldn't be done on command arguments *without care*. If the command arguments include any paths, each of those paths will still need to be normalised. The PR comment shows that it would break any non-path args passed in, like URLs. So we need to apply it with care – I don't have a good solution for this apart from examining the CLI args parser from Node itself (which is the same as qode) and maybe doing something with that – but it seems excessive for this script. So otherwise, we can just normalise the first arg, which we know is a path, and leave the rest as-is (as if they are ever introduced, it's because the user has edited the template, and at that point, it's their problem to solve!).

jonasgrunert replied:
I checked it on my machine and it works as it is now. But if you feel more comfortable normalizing all args, we can change it.

shirakaba replied:
Oh, interesting. Maybe `qode`/`node` normalises the args under-the-hood anyway, so for this particular executable we don't have to normalise them before passing them. Brilliant. If it works with macOS paths on your Windows machine without being normalised by the livereload script, then I have no further concerns!

jonasgrunert replied:
Yeah, it seems like it resolves under the hood. I checked other projects and I always read files using "/" as path seperator without any problems,

shirakaba commented:
Thank you for the PR! I'll add you to the contributors on `svelte-nodegui` itself later today.

shirakaba commented:
Added to Svelte NodeGUI contributors in PR: https://github.com/nodegui/svelte-nodegui/pull/62
